File: crypto-bot/lib/PricePredictor.php
<?php
class PricePredictor {
    public static function analyze(string $coinId = 'bitcoin'): array {
        $data = CoinGeckoApi::getHistoricalData($coinId, 90);
        $prices = array_map(fn($p) => $p[1], $data['prices']);

        $currentPrice = $prices[count($prices) - 1];
        $priceChange7d = (($prices[count($prices) - 1] - $prices[0]) / $prices[0]) * 100;

        $rsi = self::computeRsi($prices);
        $macd = self::computeMacd($prices);
        $bollinger = self::computeBollinger($prices);
        $sma50 = self::sma($prices, min(50, count($prices)));
        $sma200 = count($prices) >= 200 ? self::sma($prices, 200) : null;

        $recent20 = array_slice($prices, -20);
        $support = min($recent20);
        $resistance = max($recent20);

        $signals = [];
        $scores = [];

        if ($rsi < 30) { $signals[] = ['RSI', 'oversold', 'bullish']; $scores[] = 1; }
        elseif ($rsi > 70) { $signals[] = ['RSI', 'overbought', 'bearish']; $scores[] = -1; }
        else { $signals[] = ['RSI', 'neutral', 'neutral']; $scores[] = 0; }

        if ($macd['bullish']) { $signals[] = ['MACD', 'bullish crossover', 'bullish']; $scores[] = 1; }
        else { $signals[] = ['MACD', 'bearish crossover', 'bearish']; $scores[] = -1; }

        if ($currentPrice > $sma50) { $signals[] = ['SMA50', 'price above SMA50', 'bullish']; $scores[] = 1; }
        else { $signals[] = ['SMA50', 'price below SMA50', 'bearish']; $scores[] = -1; }

        if ($sma200 !== null) {
            if ($currentPrice > $sma200) { $signals[] = ['SMA200', 'price above SMA200', 'bullish']; $scores[] = 1; }
            else { $signals[] = ['SMA200', 'price below SMA200', 'bearish']; $scores[] = -1; }
        }

        $bbPos = $bollinger['position'];
        if ($bbPos < 20) { $signals[] = ['Bollinger', 'near lower band', 'bullish']; $scores[] = 1; }
        elseif ($bbPos > 80) { $signals[] = ['Bollinger', 'near upper band', 'bearish']; $scores[] = -1; }
        else { $signals[] = ['Bollinger', 'mid-range', 'neutral']; $scores[] = 0; }

        $avgScore = array_sum($scores) / count($scores);
        $overall = $avgScore > 0.3 ? 'BUY' : ($avgScore < -0.3 ? 'SELL' : 'HOLD');

        return [
            'current_price' => $currentPrice,
            'price_change_7d' => $priceChange7d,
            'rsi' => $rsi,
            'macd' => $macd,
            'bollinger' => $bollinger,
            'sma50' => $sma50,
            'sma200' => $sma200,
            'support' => $support,
            'resistance' => $resistance,
            'signals' => $signals,
            'overall' => $overall,
        ];
    }

    private static function computeRsi(array $prices, int $period = 14): float {
        $changes = [];
        for ($i = 1; $i < count($prices); $i++) $changes[] = $prices[$i] - $prices[$i - 1];
        $gains = array_map(fn($c) => $c > 0 ? $c : 0, $changes);
        $losses = array_map(fn($c) => $c < 0 ? -$c : 0, $changes);
        $avgGain = array_sum(array_slice($gains, 0, $period)) / $period;
        $avgLoss = array_sum(array_slice($losses, 0, $period)) / $period;
        for ($i = $period; $i < count($changes); $i++) {
            $avgGain = ($avgGain * ($period - 1) + $gains[$i]) / $period;
            $avgLoss = ($avgLoss * ($period - 1) + $losses[$i]) / $period;
        }
        if ($avgLoss == 0) return 100;
        return 100 - (100 / (1 + ($avgGain / $avgLoss)));
    }

    private static function computeMacd(array $prices): array {
        $ema12 = self::ema($prices, 12);
        $ema26 = self::ema($prices, 26);
        $macdHistory = [];
        for ($i = 0; $i < count($prices); $i++) {
            $macdHistory[] = self::ema(array_slice($prices, 0, $i + 1), 12) - self::ema(array_slice($prices, 0, $i + 1), 26);
        }
        $signal = self::ema($macdHistory, 9);
        $macdLine = $macdHistory[count($macdHistory) - 1];
        return ['macd' => $macdLine, 'signal' => $signal, 'histogram' => $macdLine - $signal, 'bullish' => $macdLine > $signal];
    }

    private static function ema(array $data, int $period): float {
        if (count($data) < $period) return array_sum($data) / count($data);
        $multiplier = 2 / ($period + 1);
        $ema = array_sum(array_slice($data, 0, $period)) / $period;
        for ($i = $period; $i < count($data); $i++) $ema = ($data[$i] - $ema) * $multiplier + $ema;
        return $ema;
    }

    private static function sma(array $data, int $period): float {
        if (count($data) < $period) $period = count($data);
        return array_sum(array_slice($data, -$period)) / $period;
    }

    private static function computeBollinger(array $prices, int $period = 20): array {
        $recent = array_slice($prices, -$period);
        $sma = array_sum($recent) / count($recent);
        $variance = array_sum(array_map(fn($p) => pow($p - $sma, 2), $recent)) / count($recent);
        $std = sqrt($variance);
        $current = $prices[count($prices) - 1];
        return [
            'upper' => $sma + 2 * $std,
            'middle' => $sma,
            'lower' => $sma - 2 * $std,
            'position' => ($std == 0) ? 50 : (($current - ($sma - 2 * $std)) / (4 * $std)) * 100,
        ];
    }
}

File: crypto-bot-py/lib/price_predictor.py
from . import coingecko_api as cg


def analyze(coin_id: str = "bitcoin") -> dict:
    data = cg.get_historical_data(coin_id, 90)
    prices = [p[1] for p in data["prices"]]

    current_price = prices[-1]
    # Use most recent ~7 days of price data (roughly 1 data point per hour = ~168 points)
    seven_days = 7 * 24
    recent_prices = prices[-seven_days:] if len(prices) > seven_days else prices
    price_change_7d = ((recent_prices[-1] - recent_prices[0]) / recent_prices[0]) * 100

    rsi = _compute_rsi(prices)
    macd = _compute_macd(prices)
    bollinger = _compute_bollinger(prices)
    sma50 = _sma(prices, min(50, len(prices)))
    sma200 = _sma(prices, 200) if len(prices) >= 200 else None

    recent20 = prices[-20:]
    support = min(recent20)
    resistance = max(recent20)

    signals = []
    scores = []

    if rsi < 30:
        signals.append(("RSI", "oversold", "bullish"))
        scores.append(1)
    elif rsi > 70:
        signals.append(("RSI", "overbought", "bearish"))
        scores.append(-1)
    else:
        signals.append(("RSI", "neutral", "neutral"))
        scores.append(0)

    if macd["bullish"]:
        signals.append(("MACD", "bullish crossover", "bullish"))
        scores.append(1)
    else:
        signals.append(("MACD", "bearish crossover", "bearish"))
        scores.append(-1)

    if current_price > sma50:
        signals.append(("SMA50", "price above SMA50", "bullish"))
        scores.append(1)
    else:
        signals.append(("SMA50", "price below SMA50", "bearish"))
        scores.append(-1)

    if sma200 is not None:
        if current_price > sma200:
            signals.append(("SMA200", "price above SMA200", "bullish"))
            scores.append(1)
        else:
            signals.append(("SMA200", "price below SMA200", "bearish"))
            scores.append(-1)

    bb_pos = bollinger["position"]
    if bb_pos < 20:
        signals.append(("Bollinger", "near lower band", "bullish"))
        scores.append(1)
    elif bb_pos > 80:
        signals.append(("Bollinger", "near upper band", "bearish"))
        scores.append(-1)
    else:
        signals.append(("Bollinger", "mid-range", "neutral"))
        scores.append(0)

    avg_score = sum(scores) / len(scores)
    if avg_score > 0.3:
        overall = "BUY"
    elif avg_score < -0.3:
        overall = "SELL"
    else:
        overall = "HOLD"

    return {
        "current_price": current_price,
        "price_change_7d": price_change_7d,
        "rsi": rsi,
        "macd": macd,
        "bollinger": bollinger,
        "sma50": sma50,
        "sma200": sma200,
        "support": support,
        "resistance": resistance,
        "signals": signals,
        "overall": overall,
    }


def _compute_rsi(prices: list[float], period: int = 14) -> float:
    changes = [prices[i] - prices[i - 1] for i in range(1, len(prices))]
    gains = [c if c > 0 else 0 for c in changes]
    losses = [-c if c < 0 else 0 for c in changes]
    avg_gain = sum(gains[:period]) / period
    avg_loss = sum(losses[:period]) / period
    for i in range(period, len(changes)):
        avg_gain = (avg_gain * (period - 1) + gains[i]) / period
        avg_loss = (avg_loss * (period - 1) + losses[i]) / period
    if avg_loss == 0:
        return 100.0
    return 100.0 - (100.0 / (1.0 + (avg_gain / avg_loss)))


def _ema(data: list[float], period: int) -> float:
    if len(data) < period:
        return sum(data) / len(data)
    multiplier = 2.0 / (period + 1)
    ema = sum(data[:period]) / period
    for i in range(period, len(data)):
        ema = (data[i] - ema) * multiplier + ema
    return ema


def _compute_macd(prices: list[float]) -> dict:
    ema12_values = []
    ema26_values = []
    for i in range(len(prices)):
        ema12_values.append(_ema(prices[: i + 1], 12))
        ema26_values.append(_ema(prices[: i + 1], 26))
    macd_history = [e12 - e26 for e12, e26 in zip(ema12_values, ema26_values)]
    signal = _ema(macd_history, 9)
    macd_line = macd_history[-1] if macd_history else 0
    return {
        "macd": macd_line,
        "signal": signal,
        "histogram": macd_line - signal,
        "bullish": macd_line > signal,
    }


def _sma(data: list[float], period: int) -> float:
    if len(data) < period:
        period = len(data)
    return sum(data[-period:]) / period


def _compute_bollinger(prices: list[float], period: int = 20) -> dict:
    recent = prices[-period:]
    sma = sum(recent) / len(recent)
    variance = sum((p - sma) ** 2 for p in recent) / len(recent)
    std = variance ** 0.5
    current = prices[-1]
    upper = sma + 2 * std
    lower = sma - 2 * std
    position = 50.0 if std == 0 else ((current - lower) / (4 * std)) * 100
    return {"upper": upper, "middle": sma, "lower": lower, "position": position}
